File: src/2025-02-24-deepclaude-rust/java/exercises/practice/two-bucket/src/main/java/TwoBucket.java
import java.util.Queue;
import java.util.LinkedList;
import java.util.Set;
import java.util.HashSet;
import java.util.ArrayList;
import java.util.List;

class TwoBucket {
    private int moves;
    private String finalBucket;
    private int otherLiters;

    TwoBucket(int bucketOneCap, int bucketTwoCap, int desiredLiters, String startBucket) {
        if (!startBucket.equals("one") && !startBucket.equals("two")) {
            throw new IllegalArgumentException("Invalid start bucket");
        }

        Queue<State> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();

        int aInitial = startBucket.equals("one") ? bucketOneCap : 0;
        int bInitial = startBucket.equals("two") ? bucketTwoCap : 0;
        queue.add(new State(aInitial, bInitial, 1));
        visited.add(aInitial + "," + bInitial);

        while (!queue.isEmpty()) {
            State current = queue.poll();

            if (current.a == desiredLiters) {
                this.moves = current.moves;
                this.finalBucket = "one";
                this.otherLiters = current.b;
                return;
            } else if (current.b == desiredLiters) {
                this.moves = current.moves;
                this.finalBucket = "two";
                this.otherLiters = current.a;
                return;
            }

            List<State> nextStates = new ArrayList<>();

            // Generate all possible next states
            nextStates.add(new State(bucketOneCap, current.b, current.moves + 1)); // Fill bucket one
            nextStates.add(new State(current.a, bucketTwoCap, current.moves + 1)); // Fill bucket two
            nextStates.add(new State(0, current.b, current.moves + 1)); // Empty bucket one
            nextStates.add(new State(current.a, 0, current.moves + 1)); // Empty bucket two

            // Pour from one to two
            int pourAmount = Math.min(current.a, bucketTwoCap - current.b);
            nextStates.add(new State(current.a - pourAmount, current.b + pourAmount, current.moves + 1));

            // Pour from two to one
            pourAmount = Math.min(current.b, bucketOneCap - current.a);
            nextStates.add(new State(current.a + pourAmount, current.b - pourAmount, current.moves + 1));

            for (State next : nextStates) {
                String key = next.a + "," + next.b;
                boolean isStartingBucketEmpty = startBucket.equals("one") ? next.a == 0 : next.b == 0;
                boolean isOtherBucketFull = startBucket.equals("one") ? next.b == bucketTwoCap : next.a == bucketOneCap;
                
                if (!isStartingBucketEmpty || !isOtherBucketFull) {
                    String key = next.a + "," + next.b;
                    if (!visited.contains(key)) {
                        visited.add(key);
                        queue.add(next);
                    }
                }
            }
        }

        throw new IllegalArgumentException("No solution exists");
    }

    int getTotalMoves() {
        return moves;
    }

    String getFinalBucket() {
        return finalBucket;
    }

    int getOtherBucket() {
        return otherLiters;
    }

    private static class State {
        int a;
        int b;
        int moves;

        State(int a, int b, int moves) {
            this.a = a;
            this.b = b;
            this.moves = moves;
        }
    }
}

File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/two-bucket/two-bucket.js
//
// This is only a SKELETON file for the 'Two Bucket' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class TwoBucket {
  static gcd(a, b) {
    while (b !== 0) {
      let t = b;
      b = a % b;
      a = t;
    }
    return a;
  }

  constructor(bucketOne, bucketTwo, goal, startBucket) {
    // Check if goal is larger than both buckets
    if (goal > Math.max(bucketOne, bucketTwo)) {
      throw new Error('Goal is larger than both buckets');
    }

    // Check if goal is achievable using GCD
    const gcd = TwoBucket.gcd(bucketOne, bucketTwo);
    if (goal % gcd !== 0) {
      throw new Error('Goal cannot be reached');
    }

    this.bucketOneCap = bucketOne;
    this.bucketTwoCap = bucketTwo;
    this.goal = goal;
    this.startBucket = startBucket;
  }

  solve() {
    const queue = [];
    const visited = new Set();

    let initialState;
    if (this.startBucket === 'one') {
      initialState = { a: this.bucketOneCap, b: 0, moves: 1 };
    } else {
      initialState = { a: 0, b: this.bucketTwoCap, moves: 1 };
    }

    visited.add(`${initialState.a},${initialState.b}`);
    queue.push(initialState);

    while (queue.length > 0) {
      const current = queue.shift();

      if (current.a === this.goal) {
        return {
          moves: current.moves,
          goalBucket: 'one',
          otherBucket: current.b
        };
      }
      if (current.b === this.goal) {
        return {
          moves: current.moves,
          goalBucket: 'two',
          otherBucket: current.a
        };
      }

      // Generate next states
      const nextStates = [];

      // Fill bucket one
      if (current.a < this.bucketOneCap) {
        const newA = this.bucketOneCap;
        const newB = current.b;
        if (newA !== current.a || newB !== current.b) {
          nextStates.push({ a: newA, b: newB, moves: current.moves + 1 });
        }
      }

      // Fill bucket two
      if (current.b < this.bucketTwoCap) {
        const newB = this.bucketTwoCap;
        const newA = current.a;
        if (newB !== current.b || newA !== current.a) {
          nextStates.push({ a: newA, b: newB, moves: current.moves + 1 });
        }
      }

      // Empty bucket one
      if (current.a > 0) {
        const newA = 0;
        const newB = current.b;
        if (newA !== current.a || newB !== current.b) {
          nextStates.push({ a: newA, b: newB, moves: current.moves + 1 });
        }
      }

      // Empty bucket two
      if (current.b > 0) {
        const newB = 0;
        const newA = current.a;
        if (newB !== current.b || newA !== current.a) {
          nextStates.push({ a: newA, b: newB, moves: current.moves + 1 });
        }
      }

      // Pour from a to b
      if (current.a > 0 && current.b < this.bucketTwoCap) {
        const transfer = Math.min(current.a, this.bucketTwoCap - current.b);
        const newA = current.a - transfer;
        const newB = current.b + transfer;
        nextStates.push({ a: newA, b: newB, moves: current.moves + 1 });
      }

      // Pour from b to a
      if (current.b > 0 && current.a < this.bucketOneCap) {
        const transfer = Math.min(current.b, this.bucketOneCap - current.a);
        const newB = current.b - transfer;
        const newA = current.a + transfer;
        nextStates.push({ a: newA, b: newB, moves: current.moves + 1 });
      }

      for (const state of nextStates) {
        // Check if the state is invalid based on the starting bucket rule
        let invalid = false;
        if (this.startBucket === 'one') {
          invalid = (state.a === 0 && state.b === this.bucketTwoCap);
        } else {
          invalid = (state.b === 0 && state.a === this.bucketOneCap);
        }

        if (!invalid) {
          const key = `${state.a},${state.b}`;
          if (!visited.has(key)) {
            visited.add(key);
            queue.push(state);
          }
        }
      }
    }

    throw new Error('No solution possible');
  }
}
